### PR TITLE
Add monitor for OpenAI field changed

### PR DESCRIPTION
**Pull request by thiendsu2303**
[#123] Add OpenAPI Schema Change Detection System

## Background

- Issue: #63 

Currently, the project lacks a mechanism to detect when request or response fields from OpenAI, Gemini, or Claude APIs have changed. Without real-time detection or notifications of these changes, API integrations may silently fail, resulting in data loss or integration errors.

## Changes

- Added new `monitor/schema` package for API schema monitoring:
  - `monitor.go`: Core monitoring logic to fetch and validate OpenAPI schemas
  - [scheduler.go](cci:7://file:///Users/maichienvithien/YanoljaWorkspace/ogem/monitor/schema/scheduler.go:0:0-0:0): Periodic schema check scheduler
  - `redis_cache.go`: Redis-based schema hash caching
  - `slack_notifier.go`: Slack notification system for schema changes
  - [monitor_test.go](cci:7://file:///Users/maichienvithien/YanoljaWorkspace/ogem/monitor/schema/monitor_test.go:0:0-0:0): Comprehensive test suite

- Key features:
  - Daily checks of OpenAI's OpenAPI schema from GitHub
  - Schema validation using kin-openapi
  - JSON normalization for consistent hash comparison
  - Redis-based caching of schema hashes
  - Slack notifications for detected changes
  - Configurable check intervals

- Updated configuration:
  - Added `slack_webhook_url` to config.yaml for notifications
  - Using existing Redis connection for schema caching

## Considerations

1. Schema Validation:
   - Strict validation of OpenAPI 3.0 format
   - Required fields checking (info.version)
   - JSON structure validation

2. Hash Consistency:
   - JSON normalization before hashing
   - Consistent handling of whitespace
   - SHA-256 for hash generation

3. Testing:
   - Comprehensive test coverage for validation and hashing
   - Edge case handling (empty schemas, invalid JSON)
   - Hash consistency verification

## Remaining Tasks

1. Add monitoring for additional providers:
   - Implement Gemini API schema monitoring
   - Implement Claude API schema monitoring
2. Add metrics collection for monitoring effectiveness
3. Consider

### CODE REVIEW

File: cmd/main.go [hunk 1]
```diff
@@ -39,6 +40,16 @@ func main() {
...
 	sugar.Infow("Loaded config", "config", config)
 
+	// Initialize schema monitor
+	cache := schema.NewRedisCache(config.ValkeyEndpoint)
+	notifier := schema.NewSlackNotifier(config.SlackWebhookURL)
+	monitor := schema.NewMonitor(sugar, cache, notifier)
+	scheduler := schema.NewScheduler(monitor, sugar)
+
+	// Start schema monitor with daily checks
+	scheduler.Start(24 * time.Hour)
```
Comment (seungduk-yanolja) [inline cmd/main.go:50]: We'd better make the server stateless. Is there any other ways to make a cron job? Seems like GitHub Actions supports it.

Comment (thiendsu2303) [reply cmd/main.go:50]: Let me research on it. Thanks for your suggestion

File: cmd/main.go [hunk 2]
```diff
@@ -55,6 +63,13 @@ func main() {
 	ctx, cancel := context.WithCancel(context.Background())
 	defer cancel()
 
+	// Start schema monitor with daily checks
```
Comment (seungduk-yanolja) [inline cmd/main.go:66]: As I said in the previous comment, this is not good. It is stateful.

Comment (thiendsu2303) [reply cmd/main.go]: I just add new update for convert it to Github Action

Comment (thiendsu2303) [reply cmd/main.go]: I will create issue and PR for this one, to split PR for easier to check

Comment (thiendsu2303) [reply cmd/main.go]: Issue for this step: #123

File: cmd/.DS_Store
Comment (seungduk-yanolja) [inline cmd/.DS_Store:1]: Please remove

File: go.mod [hunk 3]
```diff
@@ -28,16 +28,29 @@ require (
 	cloud.google.com/go/compute/metadata v0.6.0 // indirect
 	cloud.google.com/go/iam v1.2.2 // indirect
 	cloud.google.com/go/longrunning v0.6.2 // indirect
+	github.com/cespare/xxhash/v2 v2.3.0 // indirect
 	github.com/davecgh/go-spew v1.1.1 // indirect
+	github.com/dgryski/go-rendezvous v0.0.0-20200823014737-9f7001d12a5f // indirect
 	github.com/felixge/httpsnoop v1.0.4 // indirect
+	github.com/getkin/kin-openapi v0.132.0 // indirect
```
Comment (seungduk-yanolja) [inline go.mod:35]:
```
	"github.com/getkin/kin-openapi/openapi3"
	"go.uber.org/zap"
```

These are not indirect but direct dependencies

File: monitor/schema/monitor.go [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+)
+
+const (
+	OpenAISchemaURL = "https://raw.githubusercontent.com/openai/openai-openapi/master/openapi.yaml"
+	GeminiSchemaURL = "https://raw.githubusercontent.com/googleapis/google-cloud-go/main/vertexai/generativelanguage/apiv1/generativelanguage_v1.swagger.json"
+	ClaudeSchemaURL = "https://raw.githubusercontent.com/anthropics/anthropic-openapi/main/openapi.yaml"
+	cacheKeyPrefix  = "schema_cache:"
+)
+
+// Provider represents an API provider
```
Comment (seungduk-yanolja) [inline monitor/schema/monitor.go:23]: Using LLMs for coding is great, and I encourage it. However, it's important to revise and refine the code after generating it.

When writing comments, avoid stating what is already obvious from the code itself. Instead, focus on providing context that an LLM—or any reader—would not be able to infer just by reading the code.

For example, imagine giving this code to an LLM and asking it to write comments. Most of the existing comments here could easily be generated because they simply restate the code logic.

Instead, include insights like why you chose these specific providers—why OpenAI, Gemini, and Claude were selected. This kind of reasoning provides valuable context that’s not immediately evident from the code.

File: monitor/schema/monitor.go [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+type Provider string
+
+const (
+	ProviderOpenAI Provider = "openai"
+	ProviderGemini Provider = "gemini"
```
Comment (seungduk-yanolja) [inline monitor/schema/monitor.go:28]: Gemini and Claude are not providers but model names.

File: monitor/schema/monitor.go [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+// HTTPClient interface for making HTTP requests
+type HTTPClient interface {
+	Do(req *http.Request) (*http.Response, error)
+}
+
+// Monitor handles schema monitoring for various API providers
+type Monitor struct {
+	logger     *zap.SugaredLogger
+	httpClient HTTPClient
+	cache      Cache
```
Comment (seungduk-yanolja) [inline monitor/schema/monitor.go:55]: Once again, you've made it stateful here. As I mentioned earlier, I had asked for a stateless service.

Comment (seungduk-yanolja) [reply monitor/schema/monitor.go:55]: My mistake — you're using Valkey as a cache, which is good.
However, using files or other more flexible methods might be even better, as creating a strict dependency on a specific service or library is not ideal.
Ogem currently supports in-memory cache if Valkey is unavailable so no strict dependency to Valkey yet.

File: cmd/.DS_Store
Comment (thiendsu2303) [reply cmd/.DS_Store:1]: Done

File: go.mod (hunk 3 above)
Comment (thiendsu2303) [reply go.mod]: Done

File: monitor/schema/monitor.go (hunk 5 above)
Comment (thiendsu2303) [reply monitor/schema/monitor.go]: I just changed it. Thanks for your comment. Done

File: cmd/schema-monitor/main.go [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+	// Initialize schema monitor
+	httpClient := &http.Client{Timeout: 30 * time.Second}
+	cache := schema.NewRedisCache(valkeyEndpoint)
+	notifier := schema.NewSlackNotifier(slackWebhookURL)
+	monitor := schema.NewMonitor(sugar, httpClient, cache, notifier)
+
+	// Run a single check
+	ctx := context.Background()
+	if err := monitor.CheckSchemas(ctx); err != nil {
```
Comment (yoonkon-nam) [inline cmd/schema-monitor/main.go:39]:
```suggestion
	if err := monitor.CheckSchemas(context.Background()); err != nil {
```

File: monitor/schema/monitor.go [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+		return GeminiSchemaURL
+	case ProviderAnthropic:
+		return ClaudeSchemaURL
+	default:
+		return ""
+	}
+}
+
+// HTTPClient interface for making HTTP requests
+type HTTPClient interface {
```
Comment (yoonkon-nam) [inline monitor/schema/monitor.go:47]: May I ask why do we need this?
For mock?

File: monitor/schema/scheduler.go [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+	for {
+		select {
+		case <-ctx.Done():
+			s.logger.Info("Stopping schema monitor scheduler...")
+			return ctx.Err()
+		case <-ticker.C:
+			if err := s.monitor.CheckSchemas(ctx); err != nil {
+				s.logger.Errorw("Schema check failed", "error", err)
+			}
+		}
```
Comment (yoonkon-nam) [inline monitor/schema/scheduler.go:46]: If you use github cron, isn't this method meaningless?

File: monitor/schema/monitor.go (hunk 8 above)
Comment (thiendsu2303) [reply monitor/schema/monitor.go:47]: Yes, the HTTPClient interface serves three key purposes:
- Testing: Enables mocking HTTP calls in tests
- Flexibility: Allows swapping different HTTP client implementations
- Customization: Supports adding timeouts, retries, or logging without changing monitor code

File: monitor/schema/scheduler.go (hunk 9 above)
Comment (thiendsu2303) [reply monitor/schema/scheduler.go:46]: You right, I will remove al schedule.

File: cmd/schema-monitor/main.go (hunk 7 above)
Comment (thiendsu2303) [reply cmd/schema-monitor/main.go:39]: Done

File: monitor/schema/monitor.go (hunk 6 above)
Comment (thiendsu2303) [reply monitor/schema/monitor.go:55]: Let me have a research on it

File: monitor/schema/monitor.go (hunk 4 above)
Comment (thiendsu2303) [reply monitor/schema/monitor.go]: Done. The comments now focus on explaining the "why" behind design decisions rather than just restating what the code does.

File: monitor/schema/monitor.go (hunk 6 above)
Comment (thiendsu2303) [reply monitor/schema/monitor.go]: Hi @seungduk-yanolja ,  I just use state.Manager directly for schema caching:
- Remove Redis cache and memory cache adapters
- Use state.Manager directly in Monitor
- Move MockManager to testing.go for shared test utilities
- Clean up comments and formatting

Is that match with your idea

**Comment by seungduk-yanolja** [comment]
Closing: OpenAI field monitoring included in our comprehensive monitoring implementation
